tests/storage/merge_test.py

```diff
@@ -522,28 +526,6 @@ def test_reduce_chunked_internal(self):
                 ('reduce', (base_vol.optimal_size(),), {}),
             ]
 
-    def test_reduce_chunked_leaf(self):
```

Comment:
The assumption that we don't need to reduce after merge is based on wrong assumption
about creating volume with one chunk. If you create the volume with initial_size=capacity the volume will be reduced after merge.

Changing merge logic because because allocation bug was fixed is not good. We want to keep the existing behavior and fix the allocation bug.

Reply:
`If you create the volume with initial_size=capacity the volume will be reduced after merge.` Unfortunately, that's not what I observe in the test.

The test creates block volumes using `calculate_volume_alloc_size`, which results in initial_size=capacity.
https://github.com/aesteve-rh/vdsm/blob/aesteve/small-volumes-blocksd/lib/vdsm/storage/blockVolume.py#L333

On finalize, we obtain the optimal_size and the actual_size (in this case, actual_size=capacity):

Optimal_size:
In this test we create COW volumes, so we obtain the optimal size in `optimal_cow_size`:
https://github.com/aesteve-rh/vdsm/blob/aesteve/small-volumes-blocksd/lib/vdsm/storage/blockVolume.py#L421
In this case what prevails for the optimal size calculation is this line:
https://github.com/aesteve-rh/vdsm/blob/aesteve/small-volumes-blocksd/lib/vdsm/storage/blockVolume.py#L448
Which results in `optimal_size = int(capacity * 1.1)` <- this is not in the code, but is closely equivalent to lines 448 and 449 (it also aligns the result to the extent size).

Therefore, optimal_size is always bigger than actual_size (because `int(capacity * 1.1) > capacity`), and the volume is never reduced:
https://github.com/aesteve-rh/vdsm/blob/aesteve/small-volumes-blocksd/lib/vdsm/storage/merge.py#L286
Few actual examples
- 2 GiB volumes: 2.2 GiB optimal_size > 2 GiB actual_size.
- 1 GiB volumes: 1.1 GiB optimal_size > 1 GiB actual_size.
- 1 MiB volumes: 128 MiB optimal_size = 128 MiB actual_size

Reply:
On the other hand, for sizes bigger than a chunk:

`calculate_volume_alloc_size` will give an initial_size=chunk_size, for any chosen capacity in the test.

In `optimal_cow_size` the most relevant part is this one:
https://github.com/aesteve-rh/vdsm/blob/aesteve/small-volumes-blocksd/lib/vdsm/storage/blockVolume.py#L445
Which comes from `required_size (current data offset) + chunk_size` (aligned to the next extent_size).

Thus, for all these cases, we always obtain the same result:
- 3 GiB volumes (or any volume > 2.5 GiB): 2.625 GiB optimal_size > 2.5 GiB chunk size

And volume is never reduced.

Reply:
Testing 2G volume is not interesting, real disk are much bigger - changing the test to test 10g volume.

It is possible that this test used small volume size, and relied on the wrong allocation (2.5g) to test the reduce, and now when we create 2g (or 2.2) we don't need to reduce.

Reply:
As I tried to expose in this comment: https://github.com/oVirt/vdsm/pull/293#discussion_r941206435

Any volume size bigger than a chunk will not get reduced, since:
- vdsm allocates one chunk -> `actual_size = chunk_size`
- optimal_size will be `chunk_size + offset` for leaf volumes
- optimal_size > actual_size

It definitively relied in the over-allocation to trigger the reduce. The test only creates the volumes and invokes `finalize()`, so if `calculate_volume_alloc_size` does not allocate more than the `optimal_size` it just never gets reduced.

Reply:
One way to make it pass is forcing over-allocation through fakery, e.g.:
```python
        monkeypatch.setattr(
            blockVolume.BlockVolumeManifest,
            "calculate_volume_alloc_size",
            lambda *x: 4 * GiB)
```
I guess we could also write and then free some space in the volume, so that it can be reduced.

But probably this test does not make sense anymore and can be removed.
- Sparse leaf volumes will never be over-allocated with the bug-fix in this branch, so will never reduce.
- Preallocated volumes are not allowed to be reduced.

Reply:
Fixed with https://github.com/oVirt/vdsm/pull/316